### ISSUE TITLE
Make fork := true The Default

### ISSUE THREAD

**Issue by AugustNagro** [sbt/sbt#6413]
The JVM starts up very fast these days, and there are many bugs possible from not forking. For example, I just spent 30 minutes trying to figure out why a Vertx endpoint was not being updated.

**Comment by eed3si9n** [comment]
@AugustNagro Thanks for the report.

This is an interesting one. I think my opinion leans towards fork-by-default as well because I prefer things to be correct-by-default rather than fast-by-default, but at the same time for the _most_ part, in-process emulation works well, and with the turbo mode's layered ClassLoader the startup is in the order of 10s of milliseconds whereas ClassLoading scala-library and ScalaTest alone probably would take 100s of milliseconds, and many more seconds to JIT.

**Comment by AugustNagro** [comment]
I agree, correctness is the most important thing.

I switched everything I'm working on to fork := true and did not notice any
noticable start up difference.

In fact, I wonder if not forking could cause perf degredation due to
profile pollution.

I am privileged to have a good computer, but if you or someone has an old
macbook we could run a measurement with JDK 8/11.

**Comment by eatkins** [comment]
I strongly disagree with this proposal. Forking adds enormous overhead for even simple projects. It's worse the more dependencies you have: https://github.com/eatkins/scala-build-watch-performance.

**Comment by AugustNagro** [comment]
Nice to see some measurement @eatkins , thanks for sharing.

I've realized what I want is essentially https://github.com/spray/sbt-revolver

But it would still be good if `turbo`-mode `run` was reliable. One solution would be using Java's Module System, ala [layrry](https://github.com/moditect/layrry) or by [ModuleLayer](https://docs.oracle.com/javase/9/docs/api/java/lang/ModuleLayer.html) directly.

Too keep Java 8 compatibility this improvement could be added as a [multi-release jar](https://www.baeldung.com/java-multi-release-jar).

**Comment by eatkins** [comment]
Just to be clear, turbo is not the default in sbt. The default has a layer for the scala library and reflect jars that is cached between runs. This has been the default in sbt for > 10 years. There is minimal downside to caching the scala library _except_ if your run or test task uses the global future execution context in which case it is possible to leak memory/threads through the execution context.

Turbo mode adds caching of the library jars. This can work well provided that the libraries do not have persistent global state that cannot be cleaned up _and_ if classes in the library layer do not need to load classes from the application classes (basically target/**/*.class files) via reflection. Most of the time this is true so turbo would be a good default for most users. It does break some builds, however, so we never considered actually making it the default.

The drawback to every in-process classloader strategy is that if your test or run code leaks resources they will eventually cause sbt to run out of memory and crash. Usually it is easy to identify the source of leaks with jmap but there are some leaks that may not be fixable without extensive application refactoring. Fork is a good approach here. Fork is also a good approach if you are testing JNI code because a segfault can crash the main sbt application.

I'm not sure if the jdk 9 module system could be used to enhance turbo mode. At a glance, it is not clear to me what the advantages are compared to the manual classloader layering approach we are already using. The sbt classloader layering strategy has been battle-tested for the sbt use case and switching to an entirely different approach seems as likely to introduce new problems as it is to fix any existing ones.

**Comment by AugustNagro** [comment]
> Just to be clear, turbo is not the default in sbt

Ok, I assumed `turbo` was the default.

> The drawback to every in-process classloader strategy is that if your test or run code leaks resources they will eventually cause sbt to run out of memory and crash.

It is worse than that; try `run` on a Vertx project. The first time, it will work fine. End the run with CMD+C, type `run` again, and even if you make changes the old version will still be running. And then there is the memory leak problem which you mention (and I experience for many different Scala projects).

You're right that Modules are not going to solve the problem. They make configuration reliable but not the runtime.

I still think that `fork := true` default is viable, if we agree startup is the only problem. Look at [drip](https://github.com/ninjudd/drip#how-does-it-work) for example:

> Drip is a launcher for the Java Virtual Machine that provides much faster startup times than the java command

> Unlike other tools intended to solve the JVM startup problem (e.g. Nailgun, Cake), Drip does not use a persistent JVM. There are many pitfalls to using a persistent JVM, which we discovered while working on the Cake build tool for Clojure. The main problem is that the state of the persistent JVM gets dirty over time, producing strange errors and requiring liberal use of cake kill whenever any error is encountered, just in case dirty state is the cause.

> Instead of going down this road, Drip uses a different strategy. It keeps a fresh JVM spun up in reserve with the correct classpath and other JVM options so you can quickly connect and use it when needed, then throw it away. Drip hashes the JVM options and stores information about how to connect to the JVM in a directory with the hash value as its name.

**Comment by eatkins** [comment]
I agree that setting fork to true is necessary on some projects. My issue is that setting fork true will make the ux worse for the vast majority of builds. Yes, some users, unfortunately including you, may spend some time having to track down why their build is having problems and have to change the setting. But that time for individual users pales in comparison to how much time is wasted restarting the jvm between test runs. In my current watch session, I have rebuilt and ran my project 910 times with turbo mode on. If fork was the default, sbt would have spent at least 30 extra minutes reloading classes and probably about 15 extra minutes if I wasn't using turbo mode. Tens of thousands of people use sbt daily so changing to fork by default would have a significant negative impact on build performance.

I am aware of drip and considered adapting that kind of model two years ago but it doesn't work that well for the sbt use case because it's not just startup time, it's also the JIT time as @eed3si9n mentioned. There is no way to accurately pre-JIT the necessary classes. Adding something like drip would not provide enough of a performance benefit to justify the complexity of adding it as an alternative to fork.

It's easy to argue that correctness is always the most important consideration. Nobody ever got fired for buying IBM. But sometimes other considerations are actually more important and I personally feel that this is one of them. We can agree to disagree on this but you are very unlikely to persuade me that changing the default is a good idea.

**Comment by AugustNagro** [comment]
It's easy for me to say, as someone who will not implement it, but I think Drip or a similar approach will be very close to current performance, if not faster.

I've written many JMH microbenchmarks where profile pollution absolutely kills the JIT. It's honestly a rampant problem for C2  (something I've noticed Graal, with it's isolates, does much better). I have no doubt SBT is being knee-capped by it.

For the problem of C2 warmup, just take Drip's reserve VM and run a few benchmarks exercising the Scala stdlib while it's waiting. It could even be a Dogecoin miner that contributes towards SBT maintenance cost. (kidding)

At least, the idea should not be discarded until it's tried, since it would certainly improve stability & reliability.

**Comment by eatkins** [comment]
> since it would certainly improve stability & reliability.

This is a bold claim that my experience does not validate. At any rate, I am not going to work on this either so we can stop this discussion until someone comes along who is actually interested in implementing it.

**Comment by adpi2** [comment]
Due to https://github.com/sbt/sbt/pull/6665 I think we should re-consider making `fork := true` the default.

IMO `sys.exit` is not a niche feature and it should work by default. It is used in many Scala projects (I found it  in [27K files](https://github.com/search?q=sys.exit+language%3AScala&type=Code) in the 200K Scala projects in github). That means a lot of people will have to figure out why sbt server is always being shut down.

**Comment by eatkins** [comment]
To play devil's advocate, this proposal is that we make the edit <-> compile <-> test loop significantly worse by default for the > 85% of public projects that don't use sys.exit. I would propose instead adding a feature that adds a micro library/plugin with a custom exit api that throws an exception rather than exits when run inside of sbt. The api would be something like:
build.sbt: 
```scala
  useSbtExit := true
```
UserCode.scala:
```scala
  import sbt.os.Exit
  def main(args: String): Unit = {
     ...
     sbt.os.Exit(1)
  }
```
sbt.os.Exit.scala:
```scala
  package sbt.os
  object Exit {
    class ProcessExit(code: Int) extends Throwable
    def exit(i: Int): Unit =
      if (System.getProperty("sbt.os.exit", "true") == "true") System.exit(code)
      else throw ProcessExit(code)
  } 
```

Projects that work with `fork := false` are more pleasant to work with and generally will have better overall properties than projects that don't. sbt is noticeably snappier with `fork := false`. You can get feedback in less than a second, particularly if you use `turbo := true`. This allows a flow state that for me is not possible when I have to wait more than a few seconds to see the results of my changes. If the project doesn't work with `fork := false` then it is likely that there are other systemic issues that I would like to address. Maybe I'm using a library that relies on mutable global state. Or maybe I'm leaking threads somewhere. It is an indicator that something is suboptimal about my project and I should fix it. In some sense I think that programs that run with `fork := false` are more correct than programs that don't. Robust code should be able to run in multiple environments.

In terms of developer workflow, `fork := false`, when it works (which is most of the time), is clearly optimal. Changing the default makes the default developer feedback loop in sbt suboptimal. To me this is the most important property of the system. Rather than suggesting that people who are using what is already a problematic feature of the jvm change to something that is safer and that actually works with the design of sbt, this suggestion is to change sbt in a way that makes it worse for many if not most projects. It is trading a one time cost (the migration off of System.exit) for a recurring cost (a minimum increase of latency of roughly a second for test and run see https://github.com/eatkins/scala-build-watch-performance) .

Sometimes the user does need to change and software does not need to always accommodate them.

**Comment by SethTisue** [comment]
I agree with Ethan.

**Comment by AugustNagro** [comment]
> It is an indicator that something is suboptimal about my project and I should fix it
> Sometimes the user does need to change and software does not need to always accommodate them.

By that same argument, if a project can't start up fast enough to be usable with `fork := true`, then it's bloated and slow and the developers should fix it.

But in reality, this argument is absolutist, and both properties {correctness, fast startup} can be achieved.

The developers of Closure's Cake build tool did a lot of research on this, and built a library called [Drip](https://github.com/ninjudd/drip), which SBT could use.

You might say, "it'll never be as fast as fork := false", and maybe it will be a few milliseconds slower, but not by much.

**Comment by eatkins** [comment]
> The developers of Closure's Cake build tool did a lot of research on this, and built a library called Drip, which SBT could use.
> You might say, "it'll never be as fast as fork := false", and maybe it will be a few milliseconds slower, but not by much.

Please show me a PR and then we can debate implementation strategies. For now, the topic at hand is changing the default value of the `fork` setting using the _existing_ sbt infrastructure. There is no one to implement your proposal unless you are volunteering so there is no discussion to be had there. 

I am aware that `fork := false` does cause issues for some users. But the reverse also causes issues. I have formed an educated opinion that ~~`fork := true`~~ `fork := false` is a better default even with the limitations. Again, the default does cause real problems for some. That is unfortunate and I'm not denying anyone's experience.

I have done the work to implement the classloading strategy in sbt (https://github.com/sbt/sbt/pull/4476). You may disagree with me, but my opinions have been formed by my personal experience working on this specific feature in sbt. I am not interested in debating this issue with you further until you show a commitment to actually contributing to an improved solution.

**Comment by eed3si9n** [comment]
Now that sbt 1.6.0/1.6.1 is out, I've been thinking about this again.

As @adpi2 wrote back in September

> Due to #6665 I think we should re-consider making fork := true the default.

We should reconsider `fork := true` due to #6665 but not for the reason either of us knew then. Basically `TrapExit` is named such, but it's a SecurityManager not just for trapping exits, but most of the code is actually used to register current thread and instrument any thread or ThreadGroup created within the app using SecurityManager feature, and it both awaits for the main thread to finish, and cleans up the treads afterwards.

Now that I've removed that mechanism `run` is no longer awaiting for the thread (#6767). That behavior potentially can be fixed by bringing back the current thread registration. The problem is the cleanup part.
From the observation, when I call Ctrl-C the pool gets shutdown, but the daemon threads like Akka HTTP would keep running. This means that second `run` onwards would fail because port 8080 is still in use.

Given the nature of Scala, unclosed thread usage is prevalent, and combined that with `sys.exit` potentially borking the CI process in the middle as success, at this point I think we need to fork by default.

**Comment by eatkins** [comment]
I don't think I understand the importance of the use case in #6767. There is a design pattern for blocking the main thread if you are creating a daemon process. I am assuming the existence of a class, `heavy.Server` which could be a web server, akka actor service or some other kind of daemon process. I also assume that the `heavy.Server` has a close method that will cancel all of the background threads that it creates.
Before:
```scala
import heavy.Server
object App {
  def main(args: Array[String]): Unit = {
    Server.start()
  }
}
```
After:
```scala
import heavy.Server
object App {
  def main(args: Array[String]): Unit = {
    val server = Server.start()
    try(server.synchronized(server.wait())) finally server.close()
  }
}
```
Unless something has changed, I am almost certain that when the user inputs ctrl+c that sbt will interrupt the main thread of the application if it is still running. That interrupt will cause the `server.synchronized(server.wait())` call to exit and run the finally block.

**Comment by eatkins** [comment]
I have given a lot of thought to the perspective around making `fork := true` the default and while I think I understand where you all are coming from (and I will try and articulate your position below), I still believe that it is the wrong choice. I believe that making `fork := true` the default will make sbt _worse_ for most users. That being said, I agree that there are fundamentally some problems with in process user code evaluation and that the current default is _not_ the ideal in my opinion. I have a more radical suggestion that I think would make sbt _better_ for most users with the caveat that there would be some pain associated with the implementation.

Before going into it though, I want to articulate my understanding of the issue. Some of this I've already stated above, but I am trying to distill it to its essence.

Any build tool running user code in process is a fundamentally flawed approach. Since user code can do anything, it can cause the tool to get into a state where it cannot respond to user requests because the user code has exhausted all of its resources. It also can cause bugs to appear because the user code is relying on a clean global state. There are other issues that can occur when running in process, but these two are huge and essentially unavoidable. The problem is particularly bad in the scala/sbt ecosystem because the community is highly reliant on third party libraries (including the scala standard library) to do much of anything. It is unrealistic to expect third parties to write code that conforms to the expectations of the build tool so the best approach is to fork all user code into a separate process where it can be contained and killed by sbt if needed. By doing this, the stability of sbt itself can go up dramatically because sbt can concern itself with only managing its own internal resources. As far as I know, sbt itself does not have any major memory leaks (though it has had them in the past) so theoretically sbt could run essentially indefinitely if `fork := true` were the default.

So `fork := true` should be the default. Case closed.

Except setting `fork := true` makes the experience of using sbt significantly and noticeably worse for projects that were correctly designed to work within the sbt environment. There is a startup performance penalty that is a minimum of 500ms on any hardware that I've ever used when `fork := true` is used. But it's not just startup time. Because sbt caches the code for the scala standard library by default, most of the important methods get jitted after a few iterations of test or run so that sbt actually gets _faster_ the longer it runs. When setting `fork := true`, every time a user does run or test a new jvm has to reinterpret, at minimum, the scala standard library. It kind of boggles the mind how wasteful this is given how infrequently the scala standard library is updated in a project.

The current default behavior is to cache the code from the scala standard library across iterations of `run` and `test`. Since almost all sbt projects are scala projects, this is a sensible default since the project is almost certain to use this code and thus reuse it on successive iterations. This is a nice speedup but it only applies to the scala standard library. Many projects use other heavy dependencies like akka or spark that benefit from caching their code to avoid recompilation and jit time. Turbo mode enables this code caching strategy. While implementing turbo mode, I tested a hello, world project that spun up a spark server and said hello, world from inside the spark server. With in-process evaluation but with only the scala standard library cached, it took about 8 seconds every time I ran hello world. With turbo mode, the first time took 8 seconds while each successive iteration took about 1 second (still pretty heavy but 1 second vs 8 seconds is a massive improvement). Of course I had to shutdown the server at the end of each test or else sbt would quickly run out of memory but that tradeoff is well worth it for the performance benefit.

As a thought experiment, imagine writing a REPL for scala that works by forking a new scala compiler with each new user command. This would be worse than the worst REPL I've ever used.

So consider this all in the context of #6767. I am pretty sure that if they made the change I suggested above they would not need to fork. Even if I'm missing something, I am guessing that their problem could be fixed in under 10 lines of code. If sbt changes `fork := true` to default, now there are users whose build worked fine before who will experience a performance penalty that could be anywhere from 500ms to multiple seconds every time, they run a build. When I have worked with sbt in the past, it has not been unusual for me to see the counter in `~test` to exceed 1000. Changing the default could very easily add hours or days of aggregate waiting time to individual downstream users. How do we weigh that lost time compared to the effort it would take the user to change their code to block on the main thread?

sbt could do a much better job of explaining what it does and how to work with it and not against it. With turbo mode, or if the project uses few dependencies other than the standard library, it is possible to get a very tight feedback loop in sbt in which `~test` iterations can complete in under 250ms. Few projects actually attain this though because it requires esoteric knowledge of sbt and uncommon discipline. It's not actually that hard to do though.

As far as I know, all that you need to do to make turbo mode work for most projects is to
1) do not reuse global state in test runs
2) do not use a dependency that uses runtime reflection to load classes from the project classpath (the dependency will be in an isolated classloader that is upstream from the project classpath so it can't see those classes)
3) do not leak threads, including in dependencies (this actually holds for any in-process approach including the current default)

Turbo mode is especially well suited for projects that adopt a more functional programming oriented approach where both effects and resources are carefully managed. If all you have are objects with static functions, you can cache that forever without fear. When we were testing sbt 1.3.0-RCX, we actually had turbo mode as the default on the community build and a majority of projects actually worked with it but we disabled it because each of these issues did appear in a few projects. That being said, in process evaluation may not be the right approach for CI. There is value in running tests in a relatively hermetic environment. This leads to my actual proposal:

**sbt should switch to turbo mode for interactive mode by default and `fork := true` in CI.**

This would be a huge change that would cause build breakage. But this is exactly what we have the welcome banner for. We can't trust people to read release notes. I do not read the release notes when I upgrade a lot of software and I'm not going to expect others to. The welcome banner though is a place where we have the opportunity to directly communicate with the users of sbt and tell them exactly why we made this change, how we expect them to benefit from it and what to do if it causes problems and they want to restore the old behavior.

Here is a rough draft of the welcome message. It's simultaneously informal and precise. It explains the reason for the change while warning that it could cause problems and giving a one line remediation if it does indeed cause problems.
```
Welcome to sbt 1.7.0!

sbt 1.7.0 brings significant changes designed to improve performance and stability.

In interactive mode, the `test` and `run` tasks now cache the compiled code for
library dependencies by default. For builds using many dependencies, caching this 
code provides a significant, perceivable speedup in both startup time and total
runtime of these tasks. It may also break your build under certain conditions.
Revert to the pre 1.7.0 version by adding
  `ThisBuild / turbo := false`
to your main build file.

In continuous integration, it is most reliable to evaluate the `run` and `test`
tasks in a hermetically sealed environment. sbt now forks a subprocess for these
tasks by default in a CI environment. This may cause an increase to your CI build
times. Revert to the pre 1.7.0 behavior by adding `ThisBuild / forkInCI := false` to
your main build file.

See $DOCUMENTATION_LINK for more details about these features.
```

If we made turbo mode the default for interactive mode. I would bet that at least 60% of users would see a dramatic improvement in their test run times. Many of those users would be extremely grateful for their improved experience. It is not crazy to speculate that making this change could give people days of their lives back.

The effects of making this change could have a significant positive impact on the entire scala ecosystem. If every project aspired to work with turbo mode, the quality of scala libraries would go up. It would create a rising tide. Think about how many thousands of hours daily are wasted waiting for code to compile. What might be possible if people were spending less of their time waiting to validate their code and more of their time working on their actual problems?

The fact that some builds would be broken by this change does not mean that it shouldn't be done. Switching to `fork := true` by default would also almost certainly break some exotic builds so it's not like the choice is between breaking builds or not. Builds will be broken no matter what. The nice thing though about making fork be the default in CI is that it could significantly help people migrate their builds to turbo since even if they don't read the welcome banner, they'll probably notice if their build passes in CI and not on their local machine.

This is also a teachable moment. As people migrate their builds to sbt 1.7.0, they might notice that maybe in subproject a, they got a great speedup but subproject b suddenly started getting flaky or didn't work. This may force them to learn to understand subproject b better. Maybe subproject b doesn't need to use the library that is causing the builds to be much slower than they might otherwise be. Or maybe they will decide that subproject b absolutely does need to use that library and it's worth the performance hit. Then they can make an informed decision about what code evaluation strategy they want to use. They may learn to start thinking more about resource management which may help them think of more elegant designs than they might have otherwise thought of.

I do think though that the current default is actually a kind of worst of both worlds approach. It is leaving a _lot_ of performance on the table for most builds while still causing problems for builds that require a hermetic environment given the way that they are presently written.

What if instead of complaining about how slow sbt is, people started commenting on how fast it is? I don't think people understand that sbt is actually the fastest jvm build tool when code caching works. It is also the slowest build tool when it comes to start up time. In my view, it really is both the best and the worst build tool available. I don't think its possible to fix the start up time of sbt itself, but it can accurately market itself as the fastest jvm build tool for iteration _if it is used in a very specific way_. It's unfortunate that you have to hold it correctly to use it, but it is the only jvm tool that I am aware of that can give consistent sub 250ms latency for feedback. I don't think you can achieve that by holding maven*, mill, gradle or bloop** correctly but I'd be happy to be proven wrong.

I think it's worth showing people how to hold sbt correctly even if that just means they use sbt to create a better build tool that addresses some of sbt's other problems.

* I haven't proved this for maven but I have proved it for the latter three https://github.com/eatkins/scala-build-watch-performance (unless something has changed since I last ran benchmarks)
** Bloop can give you fast compilation feedback which is useful for syntax or type errors but it can't help you with logic.

**Comment by jrudolph** [comment]
I have no strong opinion on the topic, though, I have to say, that adding `fork in run := true` to an sbt project is one of the first things I do in many cases.

That said, I don't quite understand, @eatkins,  in which situations 500ms would be a significant bottleneck for any kind of interactive development when the Scala compiler usually takes seconds for even the smallest changes... It seems that in the worst case of a basically empty Scala file this change would at most double run time, in more realistic scenarios it will add maybe 500 ms to a round-trip time of already multiple seconds.

**Comment by eatkins** [comment]
> I have no strong opinion on the topic, though, I have to say, that adding fork in run := true to an sbt project is one of the first things I do in many cases.

I'm sure that makes sense if you are coming in to a new project and want to understand it better. In most cases though, I think you are penalizing yourself needlessly.

Yes, you are right that the scala compiler is ridiculously slow although I find that it is more like 1 second not multiple seconds on simple files. I am not making a big deal about 500ms though. The difference between turbo and fork can be closer to 10 seconds as I mentioned in the spark case. Sure, I would wager that most people who are testing code that runs in spark figure out how to run their code in isolation so that they don't have to spin up a spark server when they are running tests but, when you do that, your test environment is significantly different from your production environment. All things being equal, it is better if the test code is as close as possible to the production code. There are workflows that turbo enables that I don't think anyone would do with fork because it would be so inconvenient.

I am largely making a philosophical argument that you may not agree with. Every decision regarding latency matters. Maybe the scala compiler wouldn't be so slow if the tool for building it wasn't also so slow? Every little decision to add bloat and latency might make sense in a vacuum but quickly spirals into a doom loop in which everything is slow and it doesn't feel like there is any point in making anything faster.

I will lay down my cards for anyone interested. I worked on sbt in my spare time for a few years because I thought that scala was a great language hampered by the poor performance of the dev feedback loop. I proved to myself that the faster the feedback loop, the better the code that I can write. I also proved to myself that for the types of problems I am interested in working on, the scala compiler _is_ too slow. Whatever purported benefits there are from the type system and the library ecosystem are not worth the overhead that it introduces at build time. I can write better code faster in other ecosystems. For this reason, I have personally abandoned scala even though there is much that I do like about the language. I don't expect everyone reading this thread to agree with me. You may feel that scala offers guarantees that you can't find anywhere else that are worth the overhead of a slow feedback loop. To me, it does not. I want to challenge the status quo because I think that sbt and scala are both literally wasting people's time by being so slow.

Switching to turbo mode would not fix all of the problems but it would signify a change in mindset. It would say we care about making things faster. We are going to try and help be part of the solution to the problem of the slow developer feedback loop rather than contribute to making the problem worse. We are not going to accept that a computer with 8 2 ghz cpu cores with > 100MB/s ssd can only process 1KB of incremental source code changes per second.

Fast builds are anti-fragile. Once you have achieved a fast iterative feedback loop in a project, you can preserve it indefinitely. It is simple. You just don't add anything that makes the build substantially slower. You figure out how to modularize the build so that you can work on components in isolation and not constantly trigger cascading rebuilds. Because you have gotten so used to fast feedback, you will notice immediately if things get slower. You might not notice 500ms in your projects, but I have designed a system in which I get feedback on every keystroke I type within 10-20ms. The difference between 10ms and 500ms is very noticeable to me. Let alone multiple seconds.

**Comment by eatkins** [comment]
Just so people can feel what I am talking about, I made a video demonstrating the difference between turbo, non-turbo and forked sbt performance. In this video, there is a parent process that forks an sbt server and puts sbt in `~test` mode. It then modifies a file in the build, which triggers a rebuild. It times how long it takes end to end between when the file is modified and when the test completes. The test itself just spins up an Akka server and shuts it down. The watch file is essentially `object Foo { val x = 1 }`, which the scala compiler can compile in under 50ms so this test is primarily testing the overhead of the test runner reinterpreting the scala standard library and/or akka server code depending on the mode. We run 10 warmup iterations and report statistics from 20 subsequent runs. These are the results :
 project | min (ms) | max (ms) | median (ms) | total (ms)
:------- | :------: | :------: | :-------: | :--------: 
sbt-1.4.4-turbo (3 source files) | 146 | 168 | 152 | 17184
sbt-1.4.4 (3 source files) | 484 | 532 | 493 | 28661
sbt-1.4.4-fork (3 source files) | 1689 | 2004 | 1727 | 68542

If you watch the video though, you will feel the difference between the modes. It is quite noticeable:

[![asciicast](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV.svg)](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV)

If you don't want to sit through the sbt startup time for each test, here are direct links to
1) turbo: [![asciicast](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV.svg)](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV?t=12)
2) default: [![asciicast](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV.svg)](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV?t=32)
3) fork: [![asciicast](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV.svg)](https://asciinema.org/a/IH233dXH7WaFR7mPGKs36GgfV?t=32)
Notice especially that there is a long, noticeable pause after "[info] done compiling" is printed and the test starts in fork mode.

Yes, the example is somewhat contrived and YMMV but the point is to show what is possible.

**Comment by SethTisue** [comment]
500 milliseconds here, 500 milliseconds there, pretty soon you're talking real time

fwiw, I lean towards Ethan's position on this.

it's in the spirit of giving newcomers to the language a pleasant, efficient initial experience. the things that can go wrong if you don't fork are the sorts of things that go wrong when people are doing enterprise-y things at their jobs. those folks will be fine, they are rarely doing greenfield development and they have plenty of people around them to advise them when `fork := true` is needed

**Comment by jrudolph** [comment]
True, `fork in test := true` adds a lot of complexity, so I wouldn't recommend using that by default.

`fork in run := true` is different, though. Run (for me) means spawning a fresh process with a clean environment (like other compiled languages usually also do). In that case, running in the sbt process is adding extra complexity. In that case, I rather take 500ms of extra runtime for running a more realistic setup (in my tests it's more like 300ms, but let's be pessimistic about that).

(I'm all for more speed (see sbt-revolver, sbt-optimizer, perf-map-agent, etc), but there's also diminishing returns at some point. A simpler, more realistic setup seems more useful in that case, than squeezing out the last bits of turnaround time.)

**Comment by SethTisue** [comment]
Ah, interesting, it hadn't occurred to me to treat `Test` and `Run` differently. wdyt Ethan?

**Comment by eatkins** [comment]
That is for sure a fair point about run. I'm not trying to say that we should remove fork entirely. @jrudolph obviously knows what's best for @jrudolph :) I would say that having different defaults for run and test does feel more complex. I know that I have proposed something more complex with having fork in CI, but I sort of want to back off from that idea. What I am really arguing for is making turbo the default.

This is as much a philosophical issue as it is a technical one so I am just trying to explain my position. It is an issue that I have thought about quite a lot over a period of multiple years. I appreciate everyone who has read my comments up to this point. Thank you for participating in the discussion.

Part of the problem here is indeed that we all have different use cases and different interpretation of what each task is/should be. Seth nailed it in his comment. I want new programmers to have the best possible experience and trust experienced programmers to be able to debug their issues even if that requires some external help.

I absolutely get the appeal of running in a fresh, clean environment. But, at the same time, I believe that a robust program should be embeddable in another program. To wax philosophical again, there is no such thing as a truly clean environment. The operating system state is always different on every run. The program might run on different operating systems or on different jvms. Running your code in a fresh subprocess only proves that it runs in a fresh subprocess. I have more confidence in programs that can run in an embedded sbt session than those that require a fresh process to run. For my personal definition of correctness, which I concede is not shared by everyone, programs that are embeddable in sbt are more correct than programs that require a fresh subprocess.

sbt is almost like two programs in one. There is the normal batch or shell mode and `~` mode. I extensively optimized the `~` mode of sbt. Because of this, it is not a stretch to say that sbt has been optimized for my personal use case. My use case is clearly not universal but it is the only one that sbt is specifically optimized for. The best way to use sbt is to start a session with turbo on and `~test` and iterate until you're done. If you do this and avoid using a heavy test framework that makes the scala compiler slow (scalatest for example is awful for compilation times, last time I checked at least), you can consistently get feedback from sbt within a second. This is the happy path and I want more developers to experience it but they won't so long as it isn't the default.

I also wrote the thin client for sbt (which I don't actually consider a part of the happy path) and I'm going to be honest, that was an awful project to work on. To test every ui change, I had at least a 20-25 second feedback loop that required switching to multiple terminal windows to start a fresh sbt process that had to re-resolve itself each time. The graalvm took 45 seconds each time I rebuilt the native image. Having had that experience and understanding what is possible when you can get fast feedback, I will never again work on a project that cannot reliably deliver iterative feedback in less than a second. I honestly think that I could have built the thin client in a few weeks rather than closer to 6 months if that feedback loop wasn't so horrible. It kind of made me loathe sbt by the end of it.

To echo what Seth said, I don't want new users to scala or any language to have the experience that I had building the thin client. No one should be subjected to that. New users are probably not starting off building applications that require forking a subprocess. They should be starting with "Hello, world!" which is definitely embeddable in sbt. Then, given that they are using scala, they probably should learn how to write functional code that has no global mutable state. As they develop their skills, they can write increasingly challenging problems but they still will not generally need to fork a subprocess to test their new code because they only know how to write programs that are embedded within sbt. When they first get exposed to threads or server programming, they might bork their sbt session by leaking a thread or forgetting to shutdown an embedded server. That is ok. That is a great learning experience and teaches them the importance of cleaning up their threads. We only truly learn through our mistakes. We can't protect users from having to learn the hard parts of programming and I think it is actually a disservice to try.

At the same time, it is absolutely true that if a program requires fork to run for subtle reasons and the default in sbt is to run in process, developers may have a hard time debugging the problem. I'm not sugar coating that reality. It's unfortunate. That is why I've emphasized in my proposal that we loudly announce the change as well as a way to restore the old behavior (and also of course builds that have already pinned `fork := true` would not be affected by the change). @jrudolph I'm not asking you or anyone else to change their existing workflow. Only you know what is best for you. But you know sbt well enough that you can change to `fork := true` where appropriate for you. I don't think that new programmers should be thinking about fork but I do think that they should get the best possible programming experience so that they can fall in love with programming and enjoy their work.

If I knew nothing about programming and my first exposure was say through spark for a big data project. I would think, ugh, this is not fun. It would be drudgery to do anything. Since it is my first exposure to programming in general, I might think that slow feedback cycles are essential to programming and be turned off from the entire field. I am always thinking about the user who knows nothing about programming. That is who I want to optimize for.

**Comment by altrack** [comment]
From my team's experience we had more issue than gains from having `fork := false`, and eventually converted all dozens repos to `fork := true`. Usually we had some strange startup/shutdown errors/issues that were resolved by adding `fork =: true`. Another dealbreaker is the `dotenv` plugin doesn't reload environment variables on `reload` without a fork.
I've never felt any increase in test runtime due to fork.
Therefore, I think that from beginner perspective fork true will yield a better experience, and one can set fork to false when seeking a performance increase.

**Comment by eatkins** [comment]
I do not think that I have expressed my perspective well over the years that this has been open, but I am going to take one last stab. I don't expect everyone to agree with me, but I have given this issue a _lot_ of thought over the years.

I understand the points that have been made about setting fork true by default and certainly agree that it in many cases it makes sense to change to fork := true -- particularly in a legacy project. Many years ago, back in the sbt 0.13 days, I was working on a medium sized backend project and switched all of our projects over to fork := true. I did notice that things got marginally slower, but that was better than the flakiness that I was seeing with fork := false. So I can relate to the impulse to thinking this would be a better default.

However, I have come to believe that if setting fork := true is necessary in a project that it is nearly always due to a bug or a design issue (but there are always a few exceptions to the rule of course). I suspect that the vast majority of the time the reason that it fixes a flaky project is because either there is mutable global state somewhere (possibly at the library level) or there is a resource leak (again, potentially at the library level). Yes, setting fork := true can be a convenient way to paper over the issue but it is still usually a negative indicator for the project in my opinion. I understand that some readers may disagree with me on this.

There is a significant cost to changing the default. This change will impose a real cost on many users who will not benefit in any material way. There are many people whose primary use case is `sbt ~test`. For those people who do not have issues with fork := false, changing the default will impose potentially a 1-2 seconds or more penalty on every iteration. Some may not care, but I can easily notice that difference. I have had days when my iteration count has reached over 1000, which would cause 20-40 minutes of extra time per day to my workflow. Even if this change were loudly announced, I am sure plenty of people would upgrade and not notice this change and just assume there was a performance regression and needlessly eat the performance hit. The amount of wasted time across potentially thousands or even tens of thousands of developers is staggering. My hunch, which I cannot prove, is that the collective cost of changing this default in terms of developer time wasted would be far worse than the time a few developers spend figuring out that they need to change their project to fork := false. Those who need fork := true will figure it out but I do not think this preference should be imposed on all users. This change affects automated workflows and so it has a huge blast radius.

I could be wrong in my analysis, but I also don't think this issue thread is a representative sample of the community. Very few people have commented and they are presumably those who have been frustrated by the perception that they had to waste time because sbt had a bad default for _their_ project. That is understandable and I get that those of you who have commented probably correctly believe that if `fork := true` had been the default then you would have saved a significant amount time. I'm not denying your experience. I just believe that in the big picture, this change will be a net negative and think that given that it has been the default for ~15 years, we should have much better data than has been presented in this thread to make it.

**Comment by altrack** [comment]
I have never thought about this way, but all issues I have encountered were in `run`. So probably setting `fork in run := true` by default is a good middle ground.

**Comment by eed3si9n** [comment]
Going to close this issue for now. sbt 2.x will fork `run` if sbtn is used - https://github.com/sbt/sbt/pull/8060.
